lib/python/ray/worker.py

```diff
@@ -102,6 +102,102 @@ def __del__(self):
     """Deallocate the relevant segment to avoid a memory leak."""
     ray.lib.unmap_object(self.handle, self.segmentid)
 
+class RayReusables(object):
+  """An object used to store Python variables that are shared between tasks.
+
+  Each worker process will have a single RayReusables object. This class serves
+  two purposes. First, some objects are not serializable, and so the code that
+  creates those objects must be run on the worker that uses them. This class is
+  responsible for running the code that creates those objects. Second, some of
+  these objects are expensive to create, and so they should be shared between
+  tasks. However, if a task mutates a variable that is shared between tasks,
+  then the behavior of the overall program may be nondeterministic (it could
+  depend on scheduling decisions). To fix this, if a task uses a one of these
+  shared objects, then tht shared object will be reinitialized after the task
+  finishes. Since the initialization may be expensive, the user can pass in
+  custom reinitialization code that resets the state of the shared variable to
+  the way it was after initialization. If the reinitialization code does not do
+  this, then the behavior of the overall program is undefined.
```

Comment:
Random idea that just popped into my head: Since we don't want people to abuse this, we can try enforcing it by randomly spot-checking programs. (i.e., Maybe every 10-20 or so tasks, randomly recompute the same function using an initialized and a re-initialized variable, and check that they return the same thing. If they don't, crash.) It shouldn't really affect performance, and it'd help users catch bugs as well.